### DISCUSSION TITLE
Better type inference in conditional expressions

### DISCUSSION THREAD

**Discussion by Pyrdacor** [dotnet/csharplang#6736] [Q&A]
I hope this isn't covered somewhere else, but if so please feel free to close this with a duplicate reference.

I found that the following simple code won't compile:

```cs
bool foo = false;
uint z = foo ? 1 : 2;
```

The same is true for switch expressions:

```cs
uint z = x switch
{
    _ => foo ? 1 : 2
};
```

The error is that an int can't be implicitly converted to uint. I understand that in general but here it is very clear that in any case the value is a valid uint.

Is this not possible to allow for some reason? I know I can use casts or literal suffixes like `1u`. But I would prefer to avoid that as in some cases a type might change later and it should work without it imo. And suffixes won't work for smaller types like `byte`.

The same is true for any other integer type like `byte` etc which has no implicit conversation from int. So type `long` will work of course for obvious reasons.

**Comment by bernd5** [comment]
The result type of the non-const conditional expression `foo ? 1 : 2` is `int`.

But you can write:
```csharp
using System;

bool foo = false;
uint z = foo ? 1u : 2;
```
which changes the result type to `unit`. And because of implicit constant conversion the value `2` does not need an explicit conversion - just a single cast / literal suffix needed.

If you change your code to:
```csharp
const bool foo = false;
uint z = foo ? 1 : 2;
```
Than implicit constant conversion would apply here, too.

**Reply by Pyrdacor** [reply]
Yeah I know that. The question is why the compiler can't just treat this as an uint? Every "branch" has a constant expression which is compatible with the type of the left hand side.

I know that the whole expression is not a constant expression due to the dynamic boolean, but if every possible branch is a constant, you can check this at compile time.

The compiler already checks possible constant "branches" for switch cases, so why can't it be done here as well?

**Reply by CyrusNajmabadi** [reply]
> The question is why the compiler can't just treat this as an uint? 

Compat.  We tried this at one point and it broke *lots* of stuff due to code changing meaning.  So we preserve the meaning exactly if the conditional has branches all of which are compatible with each other.

**Reply by Pyrdacor** [reply]
So as each branch is a Int32 literal, it will keep this type regardless of the left hand type?

Can't think of any circumstance were it breaks but I guess you have encountered much more use cases.

Thanks for the explanation.

**Reply by CyrusNajmabadi** [reply]
> So as each branch is a Int32 literal, it will keep this type regardless of the left hand type?

Yes.

**Reply by 333fred** [reply]
> Can't think of any circumstance were it breaks but I guess you have encountered much more use cases.

The big things are overload resolution and generic type inference. It would change what overloads are called and what types are inferred, breaking existing code.

**Reply by Pyrdacor** [reply]
While the initial answer isn't that satisfying, the comments are. Should I mark the initial answer as "the answer"?
